substrate/frame/support/procedural/src/pallet/expand/event.rs

```diff
@@ -170,5 +195,17 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
 		impl<#event_impl_gen> From<#event_ident<#event_use_gen>> for () #event_where_clause {
 			fn from(_: #event_ident<#event_use_gen>) {}
 		}
+
+		impl<#event_impl_gen> #event_ident<#event_use_gen> #event_where_clause {
+			#[allow(dead_code)]
+			#[doc(hidden)]
+			pub fn event_metadata<W: #frame_support::__private::scale_info::TypeInfo + 'static>() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
```

Comment:
isn't `W` actually `Self`? why not bounding `TypeInfo` and `'static` on the impl directly?

Reply:
I was hitting this issue in some cases [link](https://github.com/paritytech/polkadot-sdk/actions/runs/9836299016/job/27151718829?pr=4948#step:4:1987).
If there's a better solution for it, I'd like to use that instead, but honestly I have no idea how to do that

Reply:
Adding to the where clause: `Self: TypeInfo` should work:

```diff
diff --git a/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs b/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
index 9ea7e2de620..27c295a7edf 100644
--- a/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
+++ b/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
@@ -225,7 +225,7 @@ fn expand_pallet_metadata_events(
 
                quote! {
                        Some(
-                               #pallet_event::event_metadata::<#pallet_event>()
+                               #pallet_event::event_metadata()
                        )
                }
        } else {
diff --git a/substrate/frame/support/procedural/src/pallet/expand/event.rs b/substrate/frame/support/procedural/src/pallet/expand/event.rs
index 947c5c83ff7..960001f8236 100644
--- a/substrate/frame/support/procedural/src/pallet/expand/event.rs
+++ b/substrate/frame/support/procedural/src/pallet/expand/event.rs
@@ -79,6 +79,10 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
                        unreachable!("Checked by event parser")
                }
        };
+       let event_where_clause_and_type_info = super::merge_where_clauses(&[
+               &event.where_clause,
+               &syn::parse_quote! { where Self: #frame_support::__private::scale_info::TypeInfo },
+       ]);
 
        // Phantom data is added for generic event.
        if event.gen_kind.is_generic() {
@@ -196,12 +200,14 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
                        fn from(_: #event_ident<#event_use_gen>) {}
                }
 
-               impl<#event_impl_gen> #event_ident<#event_use_gen> #event_where_clause {
+               impl<#event_impl_gen> #event_ident<#event_use_gen>
+                       #event_where_clause_and_type_info
+               {
                        #[allow(dead_code)]
                        #[doc(hidden)]
-                       pub fn event_metadata<W: #frame_support::__private::scale_info::TypeInfo + 'static>() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
+                       pub fn event_metadata() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
                                #frame_support::__private::metadata_ir::PalletEventMetadataIR {
-                                       ty: #frame_support::__private::scale_info::meta_type::<W>(),
+                                       ty: #frame_support::__private::scale_info::meta_type::<Self>(),
                                        deprecation_info: #deprecation_status,
                                        deprecated_variants: #variants
                                }
```